wayland: Add support for the session management protocol

Kontrabant opened the pull request:
Add session management protocol support for saving/restoring toplevel window state across runs.

Draft pull for https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/18

Kontrabant commented on src/video/wayland/SDL_waylandwindow.c:
```diff
@@ -2537,6 +2593,17 @@ bool Wayland_CreateWindow(SDL_VideoDevice *_this, SDL_Window *window, SDL_Proper
+        }
+        if (window_id) {
+            data->window_id = SDL_strdup(window_id);
+            Wayland_CreateSession(c);
```
This just defers the creation of the session object until a window with a valid ID actually requires it. This removes the requirement that the client set the session ID property before initializing the video subsystem, as clients tend to just initialize all of their required subsystems at once during early initialization, and it may need the IO subsystem initialized to retrieve the serialized session ID string from storage.

davidedmundson replied:
Ack.

Kontrabant commented on src/video/wayland/SDL_waylandwindow.c:
```diff
@@ -1502,6 +1503,77 @@ static const struct wp_fractional_scale_v1_listener fractional_scale_listener =
+        if (session_id && *session_id == '\0') {
+            session_id = NULL;
+        }
+        data->xdg_session_v1 = xdg_session_manager_v1_get_session(data->xdg_session_manager_v1, XDG_SESSION_MANAGER_V1_REASON_LAUNCH, session_id);
```
In almost all cases though, we're creating a session due to being launched normally. The 'restore' reason seems to be somewhat ambiguous in the spec:
>A app instance is restored, for example part of a restored session, or  restored from having been temporarily terminated due to resource constraints.

My interpretation of this was that it should be used if recovering from some kind of non-fatal abnormal behavior (versus 'recover' in the case of a hard crash).

Kontrabant replied:
Looking at other client implementations, Firefox sets the restored reason when a session id is passed, while others just use launch. I guess we'll do what Firefox does here, as that should be safe.

Not that it really matters, as the existing implementations in KWin and Mutter don't seem to know what to do with this information either, and just ignore it.

Kontrabant commented on src/video/wayland/SDL_waylandwindow.c:
```diff
@@ -1502,6 +1503,77 @@ static const struct wp_fractional_scale_v1_listener fractional_scale_listener =
+
+static void handle_xdg_session_replaced(void *data, struct xdg_session_v1 *xdg_session_v1)
+{
+    // NOP
```
Yes, any existing objects associated with the session also need to be destroyed here as well, as the spec states that they become inert if this happens.

davidedmundson replied:
They should. But if they're inert it means the compositor has to ignore it, so clients should be able to just do whatever is easiest for them.

Kontrabant commented:
The restriction regarding allowing a toplevel to only be added to a session before the first commit to the underlying wl_surface is a problem, as SDL destroys and recreates the Wayland toplevel objects when hiding/showing windows, while the surface persists for the life of the SDL_Window object.

PolyMeilex commented on src/video/wayland/SDL_waylandwindow.c:
```diff
@@ -1502,6 +1503,77 @@ static const struct wp_fractional_scale_v1_listener fractional_scale_listener =
+        xdg_session_v1_add_listener(data->xdg_session_v1, &xdg_session_listener, data);
+
+        // Need to round trip as the session ID is needed now.
+        WAYLAND_wl_display_roundtrip(data->display);
```
FYI the immediate ID response guarantee was dropped from the protocol spec, so you can end up with null id after this roundtrip.

ref: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/18#note_2353465

Kontrabant replied:
Done, and updated the docs to reflect that reading this property should be deferred as late as possible.

Kontrabant committed:
wayland: Add support for the session management protocol

Add session management protocol support for saving/restoring toplevel window state across runs.

Kontrabant commented:
Marking this for 3.6 since the protocol is (finally) merged upstream, and compositors are starting to ship with support for it, so implementations can be tested.

Kontrabant committed:
wayland: Add support for the toplevel tag protocol

Uses the same ID string as the session manager protocol.

Kontrabant commented:
Looking this over and testing again today, there was nothing left to do but nitpick over the documentation, so I think this is ready to go.

I'll look it over once more tomorrow and merge if nothing else comes up.

Kontrabant commented:
A couple more touch ups, and this is good to go. Merging.
